### PR TITLE
Remove the custom smartypants code [superseded by #3666]

### PR DESCRIPTION
**Pull request by mitya57**
Remove the custom smartypants code, instead rely on docutils’ ‘[smart_quotes](https://sourceforge.net/p/docutils/code/HEAD/tree/trunk/docutils/docutils/utils/smartquotes.py)’ option which is available since docutils 0.10. See #3345 for details.

This adds support for internationalization: our code supported only English quotes, while docutils code supports 27 different languages.

Add tests for English and Russian smart quotes.
Remove sphinxquotedblleft and sphinxquotedblright, should be no longer needed (@jfbu please confirm that this is fine).

### CODE REVIEW

**Comment by mitya57** [comment]
Off-topic: We now have a whole printed book at work, created with Sphinx. Thanks to you all for such a useful project :)
[Link to the photo](https://cloud.githubusercontent.com/assets/1381584/24008419/a092ba74-0a82-11e7-9d11-67feca440d03.jpg)

**Comment by jfbu** [comment]
Thanks for all this! I am all in favor, but currently it breaks French LaTeX documents because the smart quotes code introduces U+202F after ``«`` or before ``»``, and LaTeX is not configured for this Unicode character:
```
! Package inputenc Error: Unicode char   (U+202F)
(inputenc)                not set up for use with LaTeX.
```

(edited)

Current "frenchb" latex babel language module handles U+202F automatically for LuaLaTeX and XeLaTeX (meaning that it does not add extra space it it realizes it is there) but it does not for pdfLaTeX, even if the latter is configured to accept U+202F. But by default the ``«`` and ``»`` do not trigger the extra spacing done by "frenchb" so at least there will not be doubled spacing.

A 

      \DeclareUnicodeCharacter{202F}{\leavevmode\,}

or rather, better

      \DeclareUnicodeCharacter{202F}{\FBguillspace}

declaration is missing. However there is a problem for the latter choice which is that it ties to use of babel+frenchb. But I guess advanced French Sphinx LaTeX user can all overwrite it if needed.

(it is a bit complex to check "frenchb" code  due to different codepaths whether pdflatex, xelatex, lualatex as it tries to adapt spacing to the used font).

I tend to think doing per default (for pdflatex engine)

      \DeclareUnicodeCharacter{202F}{\leavevmode\,}

could be useful to other languages too. But for thoroughness Sphinx probably needs to use `\FBguillspace` rather than `\leavevmode\,` in the case of French language (not that the difference is big, but if now Sphinx is used to edit books ;-) )

edit: situation is more complicated for French because user may insert U+202F in front of high punctuation, !?;: and there we don't want it to be `\FBguillspace` which is the spacing for quotes, not same thing. But I mention this only in case  [blocking apostrophe issue](https://github.com/sphinx-doc/sphinx/pull/3562#issuecomment-287445834) gets resolved, to not rejoice too fast then.

There is [uspace](http://www.ctan.org/pkg/uspace) package which does convenient automatic declarations for various Unicode Space characters in pdflatex, but it is recent one. In future it could be worthwile for Sphinx LaTeX to use it automatically.

**Comment by jfbu** [comment]
Memo: this PR fixes #3472 !

**Comment by mitya57** [comment]
> I tend to think doing per default (for pdflatex engine)
> ```latex
>  \DeclareUnicodeCharacter{202F}{\leavevmode\,}
> ```

Added in the last commit. Thanks for the suggestion!

> Memo: this PR fixes #3472 !

Oh, it also fixes #498 (escapes like `\\'` now work), and #580.

**Comment by jfbu** [comment]
I have some bad news with following source, still using French language due to a kind of personal obsession:

    qu'en est-il de la ponctuation haute? je m'inquiète! beaucoup, même; bon on va voir: mais rien n'est sûr.

It gives this at my locale in html file:

    <p>qu ›en est-il de la ponctuation haute? je m ›inquiète! beaucoup, même; bon on va voir: mais rien n ›est sûr.</p>

and same in LaTeX (apart from the html tags of course.) Clearly the use of ``<space>›`` to match the original ``'`` which is right tick but looks straight in mono fonts as used in Emacs buffer, and serves anyhow as apostrophe on French keyboard, is wrong. (do I have something mis-configured ?)

There are other complications, in LaTeX, due to French rules, but I prefer not to comment here as it is time expensive due to needed research and plunge into frenchb code, and the above issue is blocking anyhow and not tied with LaTeX.

**Comment by jfbu** [comment]
From https://sourceforge.net/p/docutils/code/HEAD/tree/trunk/docutils/docutils/utils/smartquotes.py

                  'fr':           (u'« ',  u' »', u'‹ ', u' ›'), # with narrow no-break space

problem is that in this French sentence ``Qu'en penses-tu ?`` we really don't want the ``'`` to be interpreted as a right quote... but this is what happens in my brief testing. There must be something obvious I am missing because otherwise, the docutils smart quotes seems pretty much broken for French.

**Comment by mitya57** [comment]
You are right, it does not make any distinction between a closing quote and a mid-word apostrophe…
I have just filed a [Docutils bug](https://sourceforge.net/p/docutils/bugs/313/) for that.

So this pull request cannot be merged as-is, I will look at possible solutions…

**Review by tk0miya — APPROVED** [review]
+0: I don't have strong opinion for this. So I'm okay to merge this if French problem resolved.

**Comment by mitya57** [comment]
The bug I filed against docutils now has a [patch](https://sourceforge.net/p/docutils/bugs/313/#88f3) available. However it might take a while before it is part of a Docutils release…

So we have the following options:

1. Wait until the new Docutils release and then merge this.
2. Ship a copy of docutils code (we now have transforms support so it should be easy), drop it later.
3. Monkey-patch the docutils’ `educateQuotes` function (it’s around 100 lines of code), drop it later.
4. Keep our own code and add another configuration option to switch to Docutils code.

I like option 3 most, however I can implement any other option if you like it more. @jfbu @tk0miya What do you think?

**Comment by tk0miya** [comment]
Agreed. I also like 3rd option. We have already monkey patches for docutils at `sphinx.util.nodes` and `sphinx.util.docutils`.

**Comment by jfbu** [comment]
+1 for 3rd option.

**Comment by mitya57** [comment]
I have now updated this PR to use the 3rd option.

Now type check on Travis complains about this:

    sphinx/util/smartypants.py:81: error: No overload variant of "sub" of "Pattern" matches argument types [builtins.unicode, Any]

Where line 81 is
```python
text = apostrophe_regex.sub(smart.apostrophe, text)
```
and
```python
smart.apostrophe = u'’'
```
Does anyone know how I can fix this?

**Comment by jfbu** [comment]
From https://docs.python.org/3/library/re.html I get

> Both patterns and strings to be searched can be Unicode strings as well as 8-bit strings. However, Unicode strings and 8-bit strings cannot be mixed: that is, you cannot match a Unicode string with a byte pattern or vice-versa; similarly, when asking for a substitution, the replacement string must be of the same type as both the pattern and the search string.

thus I tried this at sphinx/util/smartypants.py#L80
```
    if smart.csquote != smart.apostrophe:
        apostrophe_regex = re.compile(ur"(?<=(\w|\d))'(?=\w)", re.UNICODE)
        text = apostrophe_regex.sub(smart.apostrophe, text)
```
and it does silence the type check, however you got the latex guy pushed to its Pythonesque extremities here, so no idea if the added "u" can break something.

**Comment by jfbu** [comment]
Hmm sorry, my `ur` causes horrible failure of `make test` in Python 3 environment. Better I say nothing more about these things I don't know...

**Comment by jfbu** [comment]
Just to signal I delete my two previous comments which were completely off the mark and if the first one contained something a bit true, it was purely accidental.

File: sphinx/util/smartypants.py
Comment (tk0miya) [inline sphinx/util/smartypants.py]: Could you add type annotation here?

**Comment by tk0miya** [comment]
>Does anyone know how I can fix this?

Please mark as `# type: ignore` even if you add type annotations.
mypy is usually confuse `str` and `unicode` around `re` library. The error might be related with the behavior.

**Comment by mitya57** [comment]
I have added type annotations. I have also added `from __future__ import unicode_literals`, with that the typecheck succeeded and `# type: ignore` was not needed.

**Comment by jfbu** [comment]
This seems to work fine. I will make remarks regarding French:
- in my brief testing the elision apostrophe as in ``Voici l'Homme`` if input as a straight quote ``'`` (U+0027) is rendered as such now, thus the problem mentioned earlier in this thread is fixed, but a better choice would be the ``’`` (U+2019).

  I understand however that the transform here is related to quotes.

- relative to quotes, and again focusing on French, ideally, the characters depend upon the nesting level, here is an example:

       Luc m’écrit : « Elle m’a dit : “Ce prétendu ‘chef-d’œuvre’ est un navet.” »

  I borrow this example from https://www.druide.com/fr/enquetes/apostrophe-et-impostrophe, sorry this is entirely in French. But that is only a remark in passing. Well, occasion for me to say that's one reason why in LaTeX macro mark-up is more powerful, because it eases up such things of detecting nesting levels, avoiding active characters.

- again for quotes, and still obsessed by French, the insertion by the (patched) DocUtils code of U+202F is a bit of a problem actually. Sure, the situation is better than using the American-English curly quotes. But the problem is with the extra U+202F: it would be the right one before the ``!?;:`` punctuation symbols, but French quotes ``«»`` are usually rendered with an about normal sized non breaking space (LaTeX's frenchb has a home-made choice, which includes some level of stretching ability). But, the U+202F seems to disappear from HTML output where it becomes a normal (breaking) space. In fact French users are better off with using directly input such as ``« `` and `` »`` (with U+00A0). This is what I do usually. The HTML will use ``&nbsp;`` and LaTeX will be with U+00A0 which may be redefined there to insert the special customized space (which has some stretch) configured by the French latex babel module. It is more complicated if we end up with U+202F in LaTeX because that's a natural thing to put in front of high punctuation. Hence problematic if it can be both associated with quotes and with punctuation. Thus, for French, I think it would be better if the quoting characters ``«»`` came with U+00A0, not with U+202F. Then it would remain non-breaking in HTML, and in LaTeX, it could be redefined to the user liking without interference with the spacing before high punctuation which could use U+202F.

Anyway, not being perfect for French does not change that it brings advantages for many languages (and French can input the correct typographical entities to start with.) Thus I am OK with this. Regarding the code, it looks good to me (I am not competent for in-depth Python reviews ...)

**Review by jfbu — APPROVED** [review]

**Comment by mitya57** [comment]
@jfbu Thanks for the thorough comments!

Re the spaces to use for French: I can replace this in my code picked from Docutils, however I consider that a temporary fork and I do not want to diverge much from Docutils there.

So I would prefer if this were fixed on Docutils side. As I know almost nothing about French typography (learning, but slowly :)), can you please file a bug for this yourself?

If this is fixed on Docutils side, I can cherry-pick the changes into our temporary fork as well.

**Comment by jfbu** [comment]
@mitya57 I have mailed directly my query to the maintainer in charge of the smart quotes module of Docutils, because it seems one needs sourceforge accreditation before opening a ticket there.

In the meantime I found these two references on the net confirming my views that a full space not a narrow space should be used with French guillemets:

http://www.graphicvertigo.com/elephant/les-regles-typographiques-de-la-ponctuation-francaise-episode-01/

http://damien.jullemier.pagesperso-orange.fr/typ/typ-espace-ponctuation.htm

It is all in French though, .... I hope Docutils will do the switch from U+202F to U+00A0, Unfortunately perhaps former dominions of the Napoleonian empire ;-) , still speaking French nowadays, have other rules, I think indeed I read once on the Wikipedia guideline pages that there were divergences for spacing (thin or normal size) relative to punctuation. But I can only, as a loyal citizen, propagate the rules of *Lexique de règles typographiques en usage à l'Imprimerie nationale, Imprimerie nationale, Paris, 1990 *.

**update** I have since gotten a hold of a copy of that book and I can confirm that this is what it says. Before ?, !, and ; a non breakable thin space, and after « and before » or before the colon : an inter-word space (non-breakable). On pages 148 and 149 of the 1990 (third) edition.

**Comment by mitya57** [comment]
Just FWIW: In docutils it is possible to use different rules for fr-FR, fr-CH, fr-CA, etc.

Also you could use docutils-develop@lists.sourceforge.net mailing list if you don’t like SF :)

**Comment by mitya57** [comment]
What else should I do to get this merged? Maybe I should monkey-patch the French fixes from docutils?

**Comment by jfbu** [comment]
@mitya57 Yes, without the French fixes we can't really merge because the necessary steps for advanced LaTeX users wishing their PDF builds to circumvent the non-adequate spacing with opening and closing quotes are a bit of a pain, and then projects will have to be modified again when the French fixes are part of Docutils. I am aware your PR anyhow creates a much better situation that using the quotes suitable for English language in texts with other languages, but getting an as satisfying as possible fix in one go, looks preferable.

File: sphinx/environment/__init__.py
Comment (tk0miya) [inline sphinx/environment/__init__.py]: Hmm... this causes `html_use_smartypants` affects to other builders. This method is called for all builders.
Is this intended? If so, it should be renamed to non `html_` prefix one.
If not, it would be better to use post_transform. It can be applied only if HTML builders.

Comment (mitya57) [reply sphinx/environment/__init__.py]: I don’t see why one may want this behavior for HTML builder and not for other builders. Actually even the existing code applied to LaTeX as well as HTML (see the `educate_quotes_latex` function which I removed).

However if we want to rename the setting (and thus break compatibility), we may as well extend it with the other options Docutils provides (i.e. add support for `use_smartypants = 'alt'`), or remove it altogether as suggested by @gmilde on [sphinx-dev group][1] (if we remove it we need to make sure smart quotes is still enabled by default).

@tk0miya What do you think? Should we remove this setting?

[1]: https://groups.google.com/d/msg/sphinx-dev/6zwDVG8UG7Q/cpdx_xVdAQAJ

File: sphinx/writers/html5.py
Comment (tk0miya) [inline sphinx/writers/html5.py:919]: It seems the `bulk_text_processor()` is used only for handling smarty pants on the subclass (SmartyPantsTranslator). After this PR, it is not needed. So it would be better to remove also from parent class.

File: sphinx/environment/__init__.py
Comment (tk0miya) [reply sphinx/environment/__init__.py]: TL;DR: +1 for removing!

Certainly, LaTeX builder applies smartypants process to the document regardless. But it is not related with `html_use_smartypants`. The process is applied even if the config is disabled.
It is only used to switch the translator component to smarty-pants'ed one. So it only affects to HTML builder and its descendants (EPUB and so on).
```
# in sphinx/builders/html.py
    def init_translator_class(self):
        # type: () -> None
        if self.translator_class is None:
            if self.config.html_experimental_html5_writer and html5_ready:
                if self.config.html_use_smartypants:
                    self.translator_class = SmartyPantsHTML5Translator
                else:
                    self.translator_class = HTML5Translator
            else:
                if self.config.html_use_smartypants:
                    self.translator_class = SmartyPantsHTMLTranslator
                else:
                    self.translator_class = HTMLTranslator
```

Hence, as I commented above, this changes the behavior of the config value.

But nobody would want to change the settings of smarty quotes by builder as you said.
So the target of smartypants setting should be widen to whole of the builders.
And it is good option to integrate the settings to docutils'.
+1 for removing `html_use_smartypants` and moving to the `smarty_quotes` of docutils'.

**Review by tk0miya — CHANGES_REQUESTED** [review]

File: sphinx/environment/__init__.py
Comment (tk0miya) [reply sphinx/environment/__init__.py]: Note:
Regarding our deprecation policy, we should show deprecation message during single major version (1.6 or 1.7) like following.
https://github.com/sphinx-doc/sphinx/blob/master/sphinx/builders/latex.py#L261

**Comment by mitya57** [comment]
@tk0miya I have deprecated the configuration option (it should be safe to just remove it in Sphinx 1.7), fixed the French quotes and removed the unused methods.

**Review by tk0miya — APPROVED** [review]

**Comment by tk0miya** [comment]
Thank you for update. It's okay for me.
I will merge this if @jfbu says okay for French fixes.

Note: we have feature freeze in this weekend. But he'll be absent until that. So I'm requesting a special treatment to this to allow to merge after the beta release.

**Comment by jfbu** [comment]
Hi @mitya57 and thanks for all your effort! 

Hi @tk0miya  I came back a bit earlier and I was going to say LGTM for merge of @mitya57 work into 1.6, but I detected a problem with this input:

```rst
(``"entry"``)
```

it gets converted by  by `make html` into
```html
(<code class="docutils literal"><span class="pre">&quot;entry&quot;</span></code>&nbsp;)
```

notice the extra space before the closing parenthesis,
and by `make latex` into

```latex
(\sphinxcode{"entry"} )
```

which has also its extra space before closing parenthesis.

There seems to be something wrong at docutils level, and at any rate it introduces behaviour which was not present before. Any idea about what might be causing this ?

If some fix is found, then I approve merge even after 1.6b1 in time for final 1.6.

edit: I forgot to say that this was using `language = 'fr'` but @mitya57 guessed it ...

**Comment by mitya57** [comment]
@jfbu Reported as https://sourceforge.net/p/docutils/bugs/317/.

**Comment by mitya57** [comment]
I have now backported another chunk of code from docutils to fix the last issue reported by @jfbu. Also rebased against the current master.

**Comment by jfbu** [comment]
Thanks for this fix from upstream docutils. Could you rebase your commits against a checkout of 1.6-release branch (https://github.com/sphinx-doc/sphinx/tree/1.6-release) rather than master ?

**Comment by mitya57** [comment]
Rebased. Do you want a new pull request, or you will merge this manually?

**Comment by tk0miya** [comment]
Could you send us a new PR, please?

**Comment by mitya57** [comment]
Done at #3666.

**Pull request closed without merging**
